### PR TITLE
Block Directory: Update search results list UI

### LINKED ISSUE #22149: Iterate the Block directory search results
Branching off from https://github.com/WordPress/gutenberg/pull/22124#pullrequestreview-406437800. 

Here's an exploration on ways to improve the design of the block directory search results in the inserter:

Screen Shot 2020-05-06 at 11 38 11

In general, I'd recommend avoiding anything lower than 13px for font size and use of italics. 

Pinging @mtias and @StevenDufresne for feedback and comments.

### LINKED ISSUE #23894: Allow for a user-triggered search in Block Directory
When I was looking for a 'card' block, there wasn't one available on my site and it triggered automatically the Block Directory search. Once the Card block was installed and used on my site, it was always showing up for the 'card' search. But I found this card now a bit limiting and I would want to see what else is in the block directory and find another one, that I previously saw in the list offered. 

For now, a user would need to experiment with the keywords to actually trigger another Block Directory search and that would be quite cumbersome.  Sooner or later the automatic triggers won't cover all use cases. 

I did explored some possible placements for the "Search in Block Directory" link/button 

### 001 Link next to the search box
SearchinBlockDirectory-001
It would be a good place for discovery. It just feels a bit crowded.

### 002 bottom of the search results
SearchinBlockDirectory-002
That's possible. Still has some issues. Mainly it's too disconnected from the search. 

Also, a link would indicate it'll take a user away and out of the block editor to a different site, and a user might be hesitant to use it for that reason. 

### 003 a button on bottom
SearchinBlockDirectory-003
This placement feels do able and reminded me on the "Browse all" button new to the canvas inserter

### 004 button area on bottom
SearchinBlockDirectory-004
It feels like it was always there, haha... if you like the idea of a user-triggered Block Directory search. 

###005 a black button on the bottom
SearchinBlockDirectory-005
That last one is just a variation on the background color to be consistent with the "Browse all" button on the other inserter. 

###006 Blue button underneath search box
SearchinBlockDirectory-006
I had second thoughts about the placement on the bottom of the search result, as people might not find it. They might not scroll all the way through the list of results. It looks a bit heavy underneath the search box, but people won't miss it there.... haha. 

Those are just a few possibilities. I like number 006 if we can make it work, and number 005 if we have to go on the bottom. But honestly none of them feel right yet. Maybe I just need to give it time? What do you think?

### PR DESCRIPTION
**Pull request by ryelle**
## Description
See https://github.com/WordPress/gutenberg/issues/22149#issuecomment-690597830 — Start implementing the new UI for block directory searches, showing block directory results even when installed blocks are found. This deviates a little from the original mockup due to issues with the BlockPreview component (requires blocks to be installed).

Fixes #22149, fixes #23894

## How has this been tested?

Search for blocks in the inserter, both installed and from [the directory.](https://wordpress.org/plugins/browse/block/) 
Make sure everything works as expected.

## Screenshots

Searching for a term that matches local blocks, block patterns, and installable blocks shows the local blocks and patterns as expected, and puts the installable blocks behind a button:

Screen Shot 2021-01-12 at 6 17 58 PM

Once clicked, the blocks show up in a list, and each is clickable to install & insert:

Screen Shot 2021-01-12 at 6 18 10 PM

Installing state:

Screen Shot 2021-01-12 at 6 18 27 PM

No local blocks but there is a block directory result, skips the button and displays the installable block(s):

Screen Shot 2021-01-12 at 6 18 51 PM

No local blocks or block directory blocks, uses the empty state:

Screen Shot 2021-01-12 at 6 20 01 PM

Error installing a block greys out the block, and shows a banner notice. The block stays greyed out on subsequent searches, but installing other blocks should still work. I would love a designer's help to make it more obvious what happened, with a more accessible design than lowering the contrast.

Screen Shot 2021-01-13 at 2 56 55 PM

### CODE REVIEW

**Comment by MichaelArestad** [comment]
> I'm not showing a block preview, because that component relies on the block already being installed :(

Is this something that we could change if we had a preview system built for the block directory (that perhaps uses generated images)?

> I dropped the description onto all views since the hovercard isn't available - we can switch that to mobile-only once the hovercard is figured out

Nice.

> The spinner makes the block author names wrap, but leaving that space for an eventual spinner made the blocks look awkwardly aligned

Good point. I think the added space is okay for now. Alternatively, we could position it absolutely over the content on top of a white gradient like this:

image

What do you think?

> We still need a place for the error notices

These, we should like do using the top notice pattern as @melchoyce [proposed here](https://github.com/WordPress/block-directory/issues/11#issuecomment-510210111). They look like this:

image

The biggest benefit is the error will be visible even if the user closes the inserter or if the inserter changes state.

### Notes
* I made some minor CSS tweaks.
* Because you didn't mention the block icons yet (instead of plugin icons), I'm assuming those haven't been implemented on the first pass. Are those still the plan?
* Can the list item descriptions be shorter? Something like three lines? I can do this with CSS if it's easier.

### Bugs

There are two that I noticed. The first is hard to spot, but sometimes there's a flash of content in the sidebar when the spinner is active during a search.

I also see a 0 results bug:

image

I would expect to see something like "No blocks found matching asdfasdfasdf"

**Comment by ryelle** [comment]
This is updated and ready for a proper review & design feedback — I've updated the description with the current status and screenshots.

File: packages/block-directory/src/components/downloadable-block-icon/index.js [hunk 1]
```diff
@@ -1,25 +1,14 @@
...
-					) }
-				/>
-			) : (
-				<BlockIcon icon={ icon } showColors />
-			) }
-		</div>
+function DownloadableBlockIcon( { icon } ) {
+	const className = 'block-directory-downloadable-block-icon';
+	return icon.match( /\.(jpeg|jpg|gif|png|svg)(?:\?.*)?$/ ) !== null ? (
+		<img className={ className } src={ icon } alt="" />
```
Comment (StevenDufresne) [inline packages/block-directory/src/components/downloadable-block-icon/index.js:9]: FYI: This was added as a result of a piece of feedback here: https://github.com/WordPress/gutenberg/pull/16524/files#r324222681

**Comment by StevenDufresne** [comment]
Awesome work on this @ryelle! It's looking really good. The code looks good.

I really like the new UX.

## A couple of notes:
I wonder if the “by” line should be on a new line. It works well first short plugin titles and 1 author. Having multiple names in the author string makes the plugin title hard to see.



**Item Loader**: I know this implementation follows the designs but the right side icon is a bit unexpected seeing that there's no button there anymore. Just adding another option for discussion.
Current | Option |
------------ | -------------
 |  |

File: packages/block-directory/src/components/downloadable-block-icon/index.js (hunk 1 above)
Comment (ryelle) [reply packages/block-directory/src/components/downloadable-block-icon/index.js:9]: The alt text, you mean? I think that feedback was about the icon labels. In this case, it sounds repetitive when using a screen reader to have "Icon of [Block Plugin]" right next to the block name, IMO the icon is functionally decorative.

Comment (StevenDufresne) [reply packages/block-directory/src/components/downloadable-block-icon/index.js:9]: Sorry wrong link: https://github.com/WordPress/gutenberg/pull/16524/files#r324268820.

But year, sounds ok.

**Comment by gziolo** [comment]
It looks like a significant improvement for Block Directory in terms of discoverability 💯 Thank you for working on it.

**Comment by shaunandrews** [comment]
image

I find the interstitial "Show X blocks" button a little strange. On one hand, I like that it clearly divides blocks you have from block you don't have yet. But, I also think its probably a little too easy to ignore.

We could default to showing any available blocks from the directory without the need to click a button:

image

That said, its probably "OK" for now, but may be worth revisiting in a new PR.

--

image

When you do click the button, you see this message. First, when there is only one result the phrase "These blocks" is a little strange. Second, it doesn't really explain what happens if you select a block. I think it could be more clear, maybe something like:

> Select a block to install and add it to your post.

--

image

When I select an available block, currently a spinner is overlaid to indicate "something is happening." We could dim the image, place the spinner over-top, hide the rating and description, and replace the byline with an "Installing..." message:

image

Once installed, the message changes to "Installed!", the new block appears in the list of installed blocks above, and the block is added to the document. I think a Snackbar notice could be helpful here as well to help explain what just happened:

image

Its worth noting that with the current behavior of the inserter sidebar, after adding a block to the document, the inserter sidebar is closed and focus is moved to the newly added block. This behavior may change (See #28191), and would likely affect the experience of installing and adding a block — for example, if the sidebar remains open, then it might make sense to remove the "Installed!" notice and block from he list of available blocks.

--

If there is an error, we can display it in the block listing directly:

image

--

Here's a few GIFs of installing a block, to help put all of the above together:

Installing a Block

Block Install Error

**Comment by ryelle** [comment]
@shaunandrews I've made the changes from your designs, you can see in the screen recording here- the first block installed has an error, which shows up as a dismissible notice and in the block card, then the second block installs successfully and shows the snackbar. I also removed the "show X blocks" step, the installable blocks list always shows up in searches now (that's not in the video).

https://user-images.githubusercontent.com/541093/105534019-e294b100-5cba-11eb-910e-41ceecb56771.mp4

**Comment by shaunandrews** [comment]
Very cool. Thanks for the update. Here's a comparison of what I seen in your video alongside the Figma design:

image

Here's a few things that stood out to me:

- If there are no blocks available, lets display a message stating that and clean up any extra spacing from the empty list of blocks.
- The "AVAILABLE TO INSTALL" heading and the text below it clash; The small, uppercase, light-gray followed by larger, darker text breaks the visual hierarchy, and just generally feels weird. In the design I avoided using the "Section Heading" styling, in favor of a more common subtitle style.
- In the design, I reduced the padding around each item in the list. I used `12px`, but we should use the related CSS variable.
- I split up the block's name and the author to make it easier to scan the list, and help avoid splitting author names into multiple lines. I emphasized the title by making it bold, and reduced the number of colors with text using the same color.
- Having the rating stars breaking up the connection between block name, author, and description makes it hard to read and scan. It also puts more emphasis on the stars than (I think) they deserve. I reduced the size of the stars, removed the review count (this could be in a tooltip), and placed them under the image. These changes go a long way towards improving the general readability of the results.
- When installing, can we hide the rating stars?
- I wonder about the need for having an error _in-line_ in the sidebar **and** the banner across the top of the editor. I guess you might close the sidebar and not see the _in-line_ notice? Seeing it in action, I'm not sure we should have both. Is reloading the page really the only way to fix things? Could we have a "Retry" button instead?
- When the block is added and the sidebar is closed, its a little jarring at first. In the design I added a Snackbar that says "XYZ block installed and added." to hopefully smooth over that sudden disappearance of the sidebar.

**Comment by ryelle** [comment]
Thanks for all the feedback, I didn't pick up on some of those changes from the screenshots 🙂  I think I've got all of it, except for the two things that we can't really do since the whole item is a button (more details below).

I've added a message when no local blocks are found, updated the "Available to install" style (I saw it like this in your first screenshot, didn't notice it was different in the others), fixed the spacing and tweaked the title & author. The rating stars have been moved & condensed a bit, and disappear while installing.

Search
Screen Shot 2021-01-25 at 5 51 11 PM

Installing
Screen Shot 2021-01-25 at 5 51 19 PM

Error
Screen Shot 2021-01-25 at 6 06 37 PM

No locally installed blocks
Screen Shot 2021-01-25 at 6 29 21 PM

> removed the review count (this could be in a tooltip)

I had to just remove it - the whole item is a button, so we can't have interactive elements like a tooltip inside. I think the number of ratings is still important, but I don't have a great idea on how to show them.

> Is reloading the page really the only way to fix things? Could we have a "Retry" button instead?

In the case of a fatal error, yeah reloading is the only way to fix it. If it's a recoverable error, we can offer a retry. [We did show a reload or retry button](https://camo.githubusercontent.com/fd0cb1f58fe3c42e4141d67ec02ef7073a6757a37f8a678c3bf9845ba9b801a4/68747470733a2f2f642e70722f692f496d456248692e706e67) - that accidentally got dropped in the redesign. Unfortunately though, same issue as above - we can't have a button in a button.

>  In the design I added a Snackbar

It's there, it shows up at the end of the video -  "Block [whatever] installed and added." It's a little delayed in the video since my local site/Chrome is slow, but it does show up right after the sidebar is closed.

Screen Shot 2021-01-25 at 5 51 44 PM

**Comment by ryelle** [comment]
@shaunandrews What are your thoughts on the issues above (that we can't have interactive elements inside the block-plugin card)?

**Comment by shaunandrews** [comment]
> I had to just remove it - the whole item is a button, so we can't have interactive elements like a tooltip inside. I think the number of ratings is still important, but I don't have a great idea on how to show them.

Perhaps we put a tooltip on the button? Something like "Install {{blockName}} - 5 stars with 392 review".

> If it's a recoverable error, we can offer a retry.

Could we just have the whole button _be_ the retry button? Maybe update the message to something like "Press to try again" or something?

**Comment by ryelle** [comment]
The whole-button tooltip works well:
Screen Shot 2021-02-26 at 2 35 42 PM

If there are no reviews I've just left off the review text:
Screen Shot 2021-02-26 at 2 38 24 PM

The button _already_ works as a retry button, so that works out. The tooltip updates in this case, too:
Screen Shot 2021-02-26 at 2 38 00 PM

**Review by shaunandrews — APPROVED** [review]
Looking good to me!

**Pull request merged**